contracts/lib.rs

```diff
@@ -469,258 +462,257 @@ pub mod prosopo {
 
             Ok(())
         }
 
-        /// Switch the `provider_account` between indexes in `self.provider_accounts`
-        fn provider_change_status(
+        /// Register a provider, their service origin and fee
+        #[ink(message)]
+        pub fn provider_register(
             &mut self,
-            provider_account: AccountId,
-            current_provider_status: GovernanceStatus,
-            new_status: GovernanceStatus,
+            service_origin: Vec<u8>,
+            fee: u32,
             payee: Payee,
-        ) {
-            if current_provider_status != new_status {
-                let current_key = ProviderState {
-                    status: current_provider_status,
-                    payee,
-                };
-                let new_key = ProviderState {
-                    status: new_status,
-                    payee,
-                };
-                // Retrieve indexes from storage mapping
-                let mut current_status_provider_accounts =
-                    self.provider_accounts.get(current_key).unwrap_or_default();
-
-                let mut new_status_provider_accounts =
-                    self.provider_accounts.get(new_key).unwrap_or_default();
-
-                // Move provider to the correct index
-                current_status_provider_accounts.remove(&provider_account);
-                new_status_provider_accounts.insert(provider_account);
-
-                // Store mapping
-                self.provider_accounts
-                    .insert(current_key, &current_status_provider_accounts);
-                self.provider_accounts
-                    .insert(new_key, &new_status_provider_accounts);
+        ) -> Result<(), Error> {
+            // this function is for registration only
+            if self.get_provider(self.env().caller()).is_ok() {
+                return err!(Error::ProviderExists);
             }
+
+            self.provider_configure(
+                Some(service_origin),
+                Some(fee),
+                Some(payee),
+                true,
+                None,
+                None,
+            )
         }
 
-        /// De-Register a provider by setting their status to Deactivated
+        /// Update an existing provider, their service origin, fee and deposit funds
         #[ink(message)]
-        pub fn provider_deregister(&mut self, provider_account: AccountId) -> Result<(), Error> {
-            let caller = self.env().caller();
-            if caller != provider_account {
-                return err!(Error::NotAuthorised);
+        #[ink(payable)]
+        pub fn provider_update(
+            &mut self,
+            service_origin: Vec<u8>,
+            fee: u32,
+            payee: Payee,
+        ) -> Result<(), Error> {
+            // this function is for updating only, not registering
+            if self.providers.get(self.env().caller()).is_none() {
+                return err!(Error::ProviderDoesNotExist);
             }
 
```

Comment:
I suppose this is where the provider diff should be carried out to determine if we need to enter `provider_configure`.